# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
pivot=pd.pivot_table(data, index=['children'], values=['debt'], aggfunc={'sum','count', 'mean'}).reset_index(drop=True)
pivot=pivot.droplevel(0,axis=1)
pivot['mean']=pivot['mean']*100
pivot['mean']=pivot['mean'].round(2)
print(pivot)

     count  mean     sum
0  14091.0  7.54  1063.0
1   4808.0  9.23   444.0
2   2052.0  9.45   194.0
3    330.0  8.18    27.0
4     41.0  9.76     4.0
5      9.0  0.00     0.0


In [27]:
pivot1=pd.pivot_table(data, index=['children'], values=['debt'], aggfunc={'sum','count', 'mean'}).droplevel(0, axis=1).reset_index()
pivot1['mean'] *= round(100, 2)
pivot1

children    count      mean     sum
0         0  14091.0  7.543822  1063.0
1         1   4808.0  9.234609   444.0
2         2   2052.0  9.454191   194.0
3         3    330.0  8.181818    27.0
4         4     41.0  9.756098     4.0
5         5      9.0  0.000000     0.0

Вывод: Исходя из процетного соотношения, можно сделать вывод, что наличие детей влияет на причину возравта кредита. Так как процент невозврата в срок у людей, у которых нет детей-7,54%.. Более высокий процент возврата кредита не в срок у людей с одним или двумя детьми: 9,23% и 9,45%. Категории людей с 3-5 детьми не будут корректно отражать данные, так как их количество меньше.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
#data['ndebt']=data['debt']
pivot=pd.pivot_table(data, index=['family_status'], values=['debt'], aggfunc={'sum','count', 'mean'}).reset_index()
pivot=pivot.droplevel(0,axis=1)
pivot['mean']=pivot['mean']*100
pivot['mean']=pivot['mean'].round(2)
print(pivot.to_string(index=False)) #не выводит столбец с индексами'
print('')
def categorize_familystatus(row):
    try:
        if row=='Не женат / не замужем' :
            return 'холост'
        elif row=='в разводе':
            return 'холост'
        elif  row=='вдовец / вдова':
            return 'холост'
        elif row=='гражданский брак':
            return 'в браке'
        elif row=='женат / замужем':
            return 'в браке'
    except:
        return 'нет категории'
data['total_family_status'] = data['family_status'].apply(categorize_familystatus)
pivot_new=pd.pivot_table(data, index=['total_family_status'], values=['debt'], aggfunc={'sum','count', 'mean'}).reset_index()
pivot_new=pivot_new.droplevel(0,axis=1)
pivot_new['mean']=pivot_new['mean']*100
pivot_new['mean']=pivot_new['mean'].round(2)
print(pivot_new.to_string(index=False))

                        count  mean   sum
Не женат / не замужем  2796.0  9.76 273.0
            в разводе  1189.0  7.06  84.0
       вдовец / вдова   951.0  6.62  63.0
     гражданский брак  4134.0  9.31 385.0
      женат / замужем 12261.0  7.56 927.0

          count  mean    sum
в браке 16395.0  8.00 1312.0
 холост  4936.0  8.51  420.0


Вывод: В данной выборке количество холостых людей (Не женат / не замужем, в разводе, вдовец / вдова) меньше-4936, чем людей, находищихся в браке,-16395. Распределив категории на две:холост и в браке, процетное соотношение невозврата кредита в срок составляет: 8,51% и 8%. Холостые люди имеют задолженность по кредиту больше, чем люди, находящиеся в браке. Но из-за несбалансировсти выборки точно опредить зависимость нельзя.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
pivot=pd.pivot_table(data, index=['total_income_category'], values=['debt'], aggfunc={'sum','count', 'mean'}).reset_index()
pivot=pivot.droplevel(0,axis=1)
pivot['mean']=pivot['mean']*100
pivot['mean']=pivot['mean'].round(2)
print(pivot.to_string(index=False))

    count  mean    sum
A    25.0  8.00    2.0
B  5014.0  7.06  354.0
C 15921.0  8.50 1353.0
D   349.0  6.02   21.0
E    22.0  9.09    2.0


Вывод: Сравнивания категории B и C, процент невозрата кредита состававлет 7,06% и 8,50% соответственно.Можно предположить, что с увеличнием дохода процент невозврата кредита в срок увеличиватеся. Остальные категории не берутся во внимание из-за равнозначенного количесво с вышеперечисленными категориями.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
pivot=pd.pivot_table(data, index=['purpose_category'], values=['debt'], aggfunc={'sum','count', 'mean'}).reset_index()
pivot=pivot.droplevel(0,axis=1)
pivot['mean']=pivot['mean']*100
pivot['mean']=pivot['mean'].round(2)
print(pivot.to_string(index=False))

                           count  mean   sum
  операции с автомобилем  4279.0  9.35 400.0
операции с недвижимостью 10751.0  7.26 780.0
   получение образования  3988.0  9.25 369.0
      проведение свадьбы  2313.0  7.91 183.0


Вывод: Меньше всего кредит берется на проведние свадьбы, процент погашения не срок-7,91%. Процентные значения для категорий "операции с автомобилем" и "получения образования" составляют 9,35% и 9,25% соответственно. Больше всего кредитов по ипотеке, не возвращают кредит в срок 7,26% людей из этой категории. Из-за несбалансированности выборки однозначно определить зависимоть невозможно. Но все категории, кроме проведение свадьбы, имеет большой процент задолженности. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Ответ:
1.Отсутвуют совершенно случайно-встреча пропуска не зависит от других переменных.
2.Отсутвует случайно- встреча пропуска зависит от других значений не своего столбца.
3.Отсутсвует неслучайно-встреча пропусков зависит от других значений. 
Исходя из классификации пропусков данных, можно сделать вывод, что многое зависит от сбора и ввода данных.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Ответ: Так как среднее значение не отражет правдивую картину. Например, з/п сотрудников  100, 100, 100, 300. Среднее значние- 150-, хотя большее количесвто получает 100. Поэтому медианное значение 100 будет отражать более точную информацию.

### Шаг 4: общий вывод.

Завершив ислледование данных о зависимостях между различными категориями и погашением кредита в срок, можно выделить следеющее:
<ul>
<li>Количество детей влияет на погащениия кредита. Процентное значение невозврата в срок у  людей с одним или двумя детьми: 9,23 и 9,45, у кого нет детей-7,54.</li>
<li>Однозначно определить зависимость между семейнйм положением и возвратом кредита невозможно из-за несбалансированости выборки. Однако процент невовзрата в срок у категории холост составляет 8,51, а у людей в браке-8.</li>  
<li>С повышнием уровня дохода увеличивается возможность задолжности по кредиту. По категорям уровня дохада В(200001–1000000) и С(50001–200000) возврат кредита в срок составляет 7,06% и 8,50%.</li>
<li>Однозначно можно сказать, что меньше всего кредит берут на проведение свадьбы (2313 людей) и задолжность по кредита составляет 7.91%. На другие цели берется больше: операции с автомобилем- 4279, операции с недвижимостью-10751, получение образования- 3988.0. Процетные значения равны:9.35, 7.26 и 9.25.</li>
</ul>
Нет сбалансииованности данных, выделено много категорий, по категориям неравное соотношение данных, поэтому однознано выделить зависимости тяжело. Рекомендуется равное соотношение по категоряим.